rust/schema.py

```diff
@@ -119,492 +187,886 @@ class Function(Declaration):
+    """
     lhs: Expr | child
     rhs: Expr | child
     op: optional[string]
 
 
-#     Range {
-#         lhs: Option<ExprId>,
-#         rhs: Option<ExprId>,
-#         range_type: RangeOp,
-#     },
-
-
+@rust.doc_test_signature("() -> ()")
 class RangeExpr(Expr):
+    """
+    A range expression. For example:
+    ```
+    let x = 1..=10;
+    let x = 1..10;
+    let x = 10..;
+    let x = ..10;
+    let x = ..=10;
+    let x = ..;
+    ```
+    """
     lhs: optional[Expr] | child
     rhs: optional[Expr] | child
     is_inclusive: predicate
 
-#     Index {
-#         base: ExprId,
-#         index: ExprId,
-#         is_assignee_expr: bool,
-#     },
-
 
+@rust.doc_test_signature("() -> ()")
 class IndexExpr(Expr):
+    """
+    An index expression. For example:
+    ```
+    list[42];
+    list[42] = 1;
+    ```
+    """
     base: Expr | child
     index: Expr | child
     is_assignee_expr: predicate
 
-#     Closure {
-#         args: Box<[PatId]>,
-#         arg_types: Box<[Option<Interned<TypeRef>>]>,
-#         ret_type: Option<Interned<TypeRef>>,
-#         body: ExprId,
-#         closure_kind: ClosureKind,
-#         capture_by: CaptureBy,
-#     },
-
 
+@rust.doc_test_signature("() -> ()")
 class ClosureExpr(Expr):
+    """
+    A closure expression. For example:
+    ```
+    |x| x + 1;
+    move |x: i32| -> i32 { x + 1 };
+    async |x: i32, y| x + y;
+     #[coroutine]
+    |x| yield x;
+     #[coroutine]
+     static |x| yield x;
+    ```
+    """
     args: list[Pat] | child
-    arg_types: list[optional[TypeRef]] | child
-    ret_type: optional[TypeRef] | child
+    arg_types: list[optional[Type]] | child
+    ret_type: optional[Type] | child
     body: Expr | child
     closure_kind: string
     is_move: predicate
-#     Tuple {
-#         exprs: Box<[ExprId]>,
-#         is_assignee_expr: bool,
-#     },
 
 
+@rust.doc_test_signature("() -> ()")
 class TupleExpr(Expr):
+    """
+    A tuple expression. For example:
+    ```
+    (1, "one");
+    (2, "two")[0] = 3;
+    ```
+    """
     exprs: list[Expr] | child
     is_assignee_expr: predicate
 
-#     Array(Array),
-
 
 class ArrayExpr(Expr):
+    """
+    An array expression. For example:
+    ```
+    [1, 2, 3];
+    [1; 10];
+    ```
+    """
     pass
-#     Literal(Literal),
-
-# ElementList { elements: Box<[ExprId]>, is_assignee_expr: bool },
 
 
+@rust.doc_test_signature("() -> ()")
 class ElementListExpr(ArrayExpr):
+    """
+    An element list expression. For example:
+    ```
+    [1, 2, 3, 4, 5];
+    [1, 2, 3, 4, 5][0] = 6;
+    ```
+    """
     elements: list[Expr] | child
     is_assignee_expr: predicate
 
-# Repeat { initializer: ExprId, repeat: ExprId },
-
 
+@rust.doc_test_signature("() -> ()")
 class RepeatExpr(ArrayExpr):
+    """
+    A repeat expression. For example:
+    ```
+    [1; 10];
+    """
     initializer: Expr | child
     repeat: Expr | child
 
 
-class LiteralExpr(Expr):
+@rust.doc_test_signature("() -> ()")
+class Literal(Expr):
```

Comment:
I'm not up-front against this, but what is the rationale behind dropping the `Expr` suffix? Isn't it a bit weird to have this one exception of all `Expr` subclasses being suffixed with `Expr`?